datafusion/physical-expr/src/physical_expr.rs

```diff
@@ -226,6 +227,15 @@ pub trait PhysicalExpr: Send + Sync + Display + Debug + PartialEq<dyn Any> {
     fn get_ordering(&self, _children: &[SortProperties]) -> SortProperties {
         SortProperties::Unordered
     }
+
+    /// The order information of the children expressions based the ordering of the expression.
```

Comment:
Actually this implementation is the reverse of the deduction from children orderings. The problem in the original issue is that, when table has ordering: `[a ASC, b ASC, c ASC]`. Assume we apply following predicate `a=round(c)` in the filter. In this case, after filtering we know that `a` and `round(c)` would be same from now on. This enables us to deduce `round(c)` is itself ordered. Which in turn, enables us to deduce `c` would be ordered.

In this case constraint (e.g ordering of the expression `round(c)`) comes from the analysis. This newly introduced API enables us to deduce given that output ordering, its children (e.g `c`) would have corresponding ordering.

With this analysis `ORDER BY c ASC` wouldn't add new `SortExec` to the plan. Since it is known that after filter, this requirement is already satisfied. Unfortunately I couldn't come up with solution without introducing a new API.

Reply:
It doesn't seem correct to me to say that `round(c)` being ordered implies `c` is ordered. For runs that have the same value for `round(c)`, the ordering of `c` can be arbitrary. For example, if the data is ordered by `c` and we project `round(c)`, followed by a `SortPreservingMergeStream` on `round(c)`, we have no guarantees that after merging the data will still be ordered by `c`, even though it was originally. 

It seems only true if the data is ordered by [`round(c), c`], as you need `c` to be the tiebreaker. But for `[round(c), b, c]` it will not be true that the output is ordered by `c` unless you know `b` is constant when `round(c)` is. So it seems to me one needs to take into account that `round(c)` and `c` are next to each other in the sort order.